### PR TITLE
Add direct endpoint api

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
- [x] revert to master, once https://github.com/nextcloud/text/pull/372 is merged
- [x] write more tests for capability

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### CODE REVIEW

File: src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
+/*
+ *
+ *  * Nextcloud Android client application
```
Comment (AndyScherzinger) [inline src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: `*  *` some comment formatting issues it seems

File: src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    @Test
+    public void testGetAll() {
+        RemoteOperationResult result = new DirectEditingObtainRemoteOperation().execute(client);
+        assertTrue(result.isSuccess());
+
+        DirectEditing directEditing = (DirectEditing) result.getSingleData();
+
+        assertEquals("Nextcloud Text", directEditing.editors.get("text").name);
+        assertEquals("text/plain", directEditing.editors.get("text").optionalMimetypes.get(0));
```
Comment (AndyScherzinger) [inline src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: maybe get this via the MimeType map? (Just a question!)

File: src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+        assertEquals("New text document", directEditing.creators.get("textdocument").name);
+        assertFalse(directEditing.creators.get("textdocument").templates);
+
+        assertEquals("New text document from template", directEditing.creators.get("textdocumenttemplate").name);
+        assertTrue(directEditing.creators.get("textdocumenttemplate").templates);
```
Comment (AndyScherzinger) [inline src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: should be a constant

File: src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        RemoteOperationResult result = new DirectEditingObtainRemoteOperation().execute(client);
+        assertTrue(result.isSuccess());
+
+        DirectEditing directEditing = (DirectEditing) result.getSingleData();
+
+        assertEquals("Nextcloud Text", directEditing.editors.get("text").name);
+        assertEquals("text/plain", directEditing.editors.get("text").optionalMimetypes.get(0));
+
+        assertEquals("New text document", directEditing.creators.get("textdocument").name);
+        assertFalse(directEditing.creators.get("textdocument").templates);
```
Comment (AndyScherzinger) [inline src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: should be a constant

File: src/main/java/com/nextcloud/android/lib/resources/directediting/DirectEditingCreateFileRemoteOperation.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    private String path;
+    private String editor;
+    private String creator;
+    private String template;
+
+    public DirectEditingCreateFileRemoteOperation(String path, String editor, String creator) {
+        this(path, editor, creator, "");
+    }
+
+    public DirectEditingCreateFileRemoteOperation(String path, String editor, String creator, String template) {
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/lib/resources/directediting/DirectEditingCreateFileRemoteOperation.java]: could be done via `lombok`

File: src/main/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainListOfTemplatesRemoteOperation.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    private static final int SYNC_READ_TIMEOUT = 40000;
+    private static final int SYNC_CONNECTION_TIMEOUT = 5000;
+    private static final String DIRECT_ENDPOINT = "/ocs/v2.php/apps/files/api/v1/directEditing/templates/";
+
+    private static final String JSON_FORMAT = "?format=json";
+
+    private String editor;
+    private String template;
+
+    public DirectEditingObtainListOfTemplatesRemoteOperation(String editor, String template) {
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainListOfTemplatesRemoteOperation.java]: could be done via `lombok`

File: src/main/java/com/nextcloud/android/lib/resources/directediting/DirectEditingOpenFileRemoteOperation.java [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    private static final int SYNC_READ_TIMEOUT = 40000;
+    private static final int SYNC_CONNECTION_TIMEOUT = 5000;
+    private static final String DIRECT_ENDPOINT = "/ocs/v2.php/apps/files/api/v1/directEditing/open";
+
+    private static final String JSON_FORMAT = "?format=json";
+
+    private String filePath;
+    private String editor;
+
+    public DirectEditingOpenFileRemoteOperation(String filePath, String editor) {
```
Comment (AndyScherzinger) [inline src/main/java/com/nextcloud/android/lib/resources/directediting/DirectEditingOpenFileRemoteOperation.java]: `lombok`

File: src/main/java/com/owncloud/android/lib/common/Creator.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+/**
+ * Creator of direct editing data model
+ */
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
```
Comment (AndyScherzinger) [inline src/main/java/com/owncloud/android/lib/common/Creator.java]: ❤️

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky looks good to me. Just some minor comments about further utilization of lombok 👍

File: src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java (hunk 2 above)
Comment (tobiasKaminsky) [reply src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: How would you do this?

File: src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java (hunk 4 above)
Comment (tobiasKaminsky) [reply src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: For testing I rather do not care, or do we want to do this there, too?

Comment (AndyScherzinger) [reply src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: Well, is this variable a constant in the regular code? Then I'd say let's refernce it there. If it isn't then I am fine leaving it as is.

Comment (tobiasKaminsky) [reply src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: No, it is not there. This is later transformed from json to a real class via gson and thus we do not need it.

File: src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java (hunk 2 above)
Comment (AndyScherzinger) [reply src/androidTest/java/com/nextcloud/android/lib/resources/directediting/DirectEditingObtainRemoteOperationTest.java]: Nevermind, the mimetype is part of the app, not the lib I think :/

**Comment by AndyScherzinger** [comment]
Looks good to me 👍

**Comment by tobiasKaminsky** [comment]
Tests fails as @juliushaertl seems to removed the templates.
Will the remain removed?

File: .drone.yml [hunk 9]
```diff
@@ -100,7 +100,7 @@ services:
       - su www-data -c "php /var/www/html/occ group:adduser users user2"
       - su www-data -c "git clone -b master https://github.com/nextcloud/activity.git /var/www/html/apps/activity/"
       - su www-data -c "php /var/www/html/occ app:enable activity"
-      - su www-data -c "git clone -b master https://github.com/nextcloud/text.git /var/www/html/apps/text/"
+      - su www-data -c "git clone -b master https://github.com/nextcloud/testing_directediting.git /var/www/html/apps/testing_directediting/"
       - su www-data -c "php /var/www/html/occ app:enable text"
```
Comment (juliushaertl) [inline .drone.yml]: the app is named testing_directediting, so it should be `occ app:enable testing_directediting`

**Comment by AndyScherzinger** [comment]
@tobiasKaminsky Spotbugs is drunk...

**Comment by tobiasKaminsky** [comment]
Hehe. Somehow yes…I will change it accordingly, once all tests are green.
(then also rebase it)

**Comment by tobiasKaminsky** [comment]
Last failing test will be resolved via https://github.com/nextcloud/text/pull/478

**Pull request merged**